File: apps/agent/src/context/weather.ts
/**
 * Open-Meteo client for weather context. Results are cached for 30 minutes
 * keyed by lat/lng rounded to 2 decimal places (~1km resolution) so nearby
 * coordinates share a cache entry.
 */

export interface Coord {
  lat: number;
  lng: number;
}

export interface Weather {
  current: { temperatureC: number; windKph: number; code: number; time: string };
  forecast: Array<{ date: string; maxC: number; minC: number; code: number }>;
  /**
   * Today-specific signals the coach uses for "should I run / get morning
   * light / bring a jacket" reasoning. Pre-computed so the LLM doesn't have
   * to scan the hourly array.
   */
  today: {
    sunrise: string;
    sunset: string;
    daylightHours: number;
    uvIndexMax: number;
    /** Peak rain hour today, or null if the whole day is below 20% chance. */
    rainChancePeak: { hour: string; probability: number } | null;
  };
}

export type WeatherFetcher = (url: string) => Promise<Response>;

export interface WeatherClient {
  get(coord: Coord): Promise<Weather | null>;
}

interface CacheEntry {
  at: number;
  value: Weather | null;
}

const TTL_MS = 30 * 60_000;

export function roundForCache(c: Coord): Coord {
  return { lat: roundTo(c.lat, 2), lng: roundTo(c.lng, 2) };
}

function roundTo(n: number, places: number): number {
  const factor = 10 ** places;
  return Math.round(n * factor) / factor;
}

function cacheKey(c: Coord): string {
  const r = roundForCache(c);
  return `${r.lat},${r.lng}`;
}

export function createWeatherClient(
  deps: {
    fetcher?: WeatherFetcher;
    now?: () => number;
    ttlMs?: number;
  } = {},
): WeatherClient {
  const fetcher = deps.fetcher ?? ((url: string) => fetch(url));
  const now = deps.now ?? (() => Date.now());
  const ttl = deps.ttlMs ?? TTL_MS;
  const cache = new Map<string, CacheEntry>();

  return {
    async get(coord: Coord): Promise<Weather | null> {
      const key = cacheKey(coord);
      const hit = cache.get(key);
      if (hit && now() - hit.at < ttl) return hit.value;

      const { lat, lng } = roundForCache(coord);
      const dailyFields =
        'temperature_2m_max,temperature_2m_min,weather_code,sunrise,sunset,uv_index_max,precipitation_probability_max,daylight_duration';
      const hourlyFields = 'precipitation_probability,uv_index,cloud_cover';
      const url = `https://api.open-meteo.com/v1/forecast?latitude=${lat}&longitude=${lng}&current=temperature_2m,weather_code,wind_speed_10m&daily=${dailyFields}&hourly=${hourlyFields}&forecast_days=7&timezone=auto`;

      try {
        const res = await fetcher(url);
        if (res.status !== 200) {
          cache.set(key, { at: now(), value: null });
          return null;
        }
        const body = (await res.json()) as OpenMeteoResponse;
        const weather: Weather = {
          current: {
            temperatureC: body.current.temperature_2m,
            windKph: body.current.wind_speed_10m,
            code: body.current.weather_code,
            time: body.current.time,
          },
          forecast: body.daily.time.map((date: string, i: number) => ({
            date,
            maxC: body.daily.temperature_2m_max[i] ?? Number.NaN,
            minC: body.daily.temperature_2m_min[i] ?? Number.NaN,
            code: body.daily.weather_code[i] ?? 0,
          })),
          today: buildTodayBlock(body),
        };
        cache.set(key, { at: now(), value: weather });
        return weather;
      } catch {
        cache.set(key, { at: now(), value: null });
        return null;
      }
    },
  };
}

interface OpenMeteoResponse {
  current: {
    time: string;
    temperature_2m: number;
    weather_code: number;
    wind_speed_10m: number;
  };
  daily: {
    time: string[];
    temperature_2m_max: number[];
    temperature_2m_min: number[];
    weather_code: number[];
    sunrise: string[];
    sunset: string[];
    uv_index_max: number[];
    precipitation_probability_max: number[];
    daylight_duration: number[];
  };
  hourly: {
    time: string[];
    precipitation_probability: number[];
    uv_index: number[];
    cloud_cover: number[];
  };
}

const RAIN_PEAK_MIN = 20;

function buildTodayBlock(body: OpenMeteoResponse): Weather['today'] {
  const todayDate = body.daily.time[0] ?? '';
  // Slice the hourly arrays to today's 24 entries.
  const todayHours: Array<{ time: string; rain: number }> = [];
  for (let i = 0; i < body.hourly.time.length; i++) {
    const t = body.hourly.time[i];
    if (!t || !t.startsWith(todayDate)) continue;
    todayHours.push({ time: t, rain: body.hourly.precipitation_probability[i] ?? 0 });
  }
  const peak = todayHours.reduce<{ time: string; rain: number } | null>(
    (best, cur) => (best === null || cur.rain > best.rain ? cur : best),
    null,
  );
  const rainChancePeak =
    peak && peak.rain >= RAIN_PEAK_MIN ? { hour: peak.time, probability: peak.rain } : null;

  return {
    sunrise: body.daily.sunrise[0] ?? '',
    sunset: body.daily.sunset[0] ?? '',
    daylightHours: roundTo((body.daily.daylight_duration[0] ?? 0) / 3600, 1),
    uvIndexMax: body.daily.uv_index_max[0] ?? 0,
    rainChancePeak,
  };
}

File: apps/agent_py/src/lifecoach_agent/context/weather.py
"""Open-Meteo client for weather context.

Results are cached for 30 minutes keyed by lat/lng rounded to 2 decimal
places (~1km resolution) so nearby coordinates share a cache entry.

Mirrors `apps/agent/src/context/weather.ts`. The TS file is the
behavioural source of truth.
"""

from __future__ import annotations

import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any

import httpx

from lifecoach_agent.prompt.build_instruction import (
    Coord,
    RainPeak,
    Weather,
    WeatherCurrent,
    WeatherForecastDay,
    WeatherToday,
)

HttpFetcher = Callable[[str], Awaitable[httpx.Response]]

_TTL_S: float = 30 * 60.0
_RAIN_PEAK_MIN: int = 20


def round_for_cache(c: Coord) -> Coord:
    """Round to 2 decimal places (~1km) so nearby coords share a cache entry."""
    return Coord(lat=_round_to(c.lat, 2), lng=_round_to(c.lng, 2))


def _round_to(n: float, places: int) -> float:
    factor = 10**places
    return float(round(n * factor) / factor)


def _cache_key(c: Coord) -> str:
    r = round_for_cache(c)
    return f"{r.lat},{r.lng}"


@dataclass
class _CacheEntry:
    at: float
    value: Weather | None


async def _default_fetcher(url: str) -> httpx.Response:
    async with httpx.AsyncClient() as client:
        return await client.get(url)


class WeatherClient:
    def __init__(
        self,
        fetcher: HttpFetcher | None = None,
        now: Callable[[], float] | None = None,
        ttl_s: float | None = None,
    ) -> None:
        self._fetcher: HttpFetcher = fetcher if fetcher is not None else _default_fetcher
        self._now: Callable[[], float] = now if now is not None else time.monotonic
        self._ttl_s: float = ttl_s if ttl_s is not None else _TTL_S
        self._cache: dict[str, _CacheEntry] = {}

    async def get(self, coord: Coord) -> Weather | None:
        key = _cache_key(coord)
        hit = self._cache.get(key)
        if hit is not None and self._now() - hit.at < self._ttl_s:
            return hit.value

        rounded = round_for_cache(coord)
        daily_fields = (
            "temperature_2m_max,temperature_2m_min,weather_code,sunrise,sunset,"
            "uv_index_max,precipitation_probability_max,daylight_duration"
        )
        hourly_fields = "precipitation_probability,uv_index,cloud_cover"
        url = (
            "https://api.open-meteo.com/v1/forecast"
            f"?latitude={rounded.lat}&longitude={rounded.lng}"
            "&current=temperature_2m,weather_code,wind_speed_10m"
            f"&daily={daily_fields}"
            f"&hourly={hourly_fields}"
            "&forecast_days=7&timezone=auto"
        )

        try:
            res = await self._fetcher(url)
            if res.status_code != 200:
                self._cache[key] = _CacheEntry(at=self._now(), value=None)
                return None
            body: dict[str, Any] = res.json()
            weather = _parse(body)
            self._cache[key] = _CacheEntry(at=self._now(), value=weather)
            return weather
        except Exception:
            self._cache[key] = _CacheEntry(at=self._now(), value=None)
            return None


def _parse(body: dict[str, Any]) -> Weather:
    current = body["current"]
    daily = body["daily"]
    forecast = [
        WeatherForecastDay(
            date=date,
            maxC=_get_or(daily.get("temperature_2m_max"), i, float("nan")),
            minC=_get_or(daily.get("temperature_2m_min"), i, float("nan")),
            code=int(_get_or(daily.get("weather_code"), i, 0)),
        )
        for i, date in enumerate(daily.get("time", []))
    ]
    return Weather(
        current=WeatherCurrent(
            temperatureC=current["temperature_2m"],
            windKph=current["wind_speed_10m"],
            code=int(current["weather_code"]),
            time=current["time"],
        ),
        forecast=forecast,
        today=_build_today_block(body),
    )


def _build_today_block(body: dict[str, Any]) -> WeatherToday:
    daily = body["daily"]
    hourly = body["hourly"]
    today_date: str = (daily.get("time") or [""])[0]

    rain_probs = hourly.get("precipitation_probability") or []
    times = hourly.get("time") or []
    today_hours: list[tuple[str, float]] = []
    for i, t in enumerate(times):
        if not t or not t.startswith(today_date):
            continue
        rain = rain_probs[i] if i < len(rain_probs) and rain_probs[i] is not None else 0
        today_hours.append((t, rain))

    peak: tuple[str, float] | None = None
    for t, rain in today_hours:
        if peak is None or rain > peak[1]:
            peak = (t, rain)

    rain_chance_peak: RainPeak | None = None
    if peak is not None and peak[1] >= _RAIN_PEAK_MIN:
        rain_chance_peak = RainPeak(hour=peak[0], probability=int(peak[1]))

    return WeatherToday(
        sunrise=_get_or(daily.get("sunrise"), 0, ""),
        sunset=_get_or(daily.get("sunset"), 0, ""),
        daylightHours=_round_to(_get_or(daily.get("daylight_duration"), 0, 0) / 3600, 1),
        uvIndexMax=_get_or(daily.get("uv_index_max"), 0, 0),
        rainChancePeak=rain_chance_peak,
    )


def _get_or(seq: list[Any] | None, i: int, default: Any) -> Any:
    if seq is None or i >= len(seq) or seq[i] is None:
        return default
    return seq[i]
